perception/traffic_light_estimator/src/node.cpp

```diff
@@ -0,0 +1,333 @@
+// Copyright 2022 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "traffic_light_estimator/node.hpp"
+
+#include <lanelet2_extension/utility/message_conversion.hpp>
+
+#include <iostream>
+#include <memory>
+#include <utility>
+#include <vector>
+
+namespace traffic_light
+{
+namespace
+{
+
+bool hasMergeLane(
+  const lanelet::ConstLanelet & lanelet_1, const lanelet::ConstLanelet & lanelet_2,
+  const lanelet::routing::RoutingGraphPtr & routing_graph_ptr)
+{
+  const auto next_lanelets_1 = routing_graph_ptr->following(lanelet_1);
+  const auto next_lanelets_2 = routing_graph_ptr->following(lanelet_2);
+
+  for (const auto & next_lanelet_1 : next_lanelets_1) {
+    for (const auto & next_lanelet_2 : next_lanelets_2) {
+      if (next_lanelet_1.id() == next_lanelet_2.id()) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool hasMergeLane(
+  const lanelet::ConstLanelets & lanelets,
+  const lanelet::routing::RoutingGraphPtr & routing_graph_ptr)
+{
+  for (size_t i = 0; i < lanelets.size(); ++i) {
+    for (size_t j = i + 1; j < lanelets.size(); ++j) {
+      const auto lanelet_1 = lanelets.at(i);
+      const auto lanelet_2 = lanelets.at(j);
+
+      if (lanelet_1.id() == lanelet_2.id()) {
+        continue;
+      }
+
+      const std::string turn_direction_1 = lanelet_1.attributeOr("turn_direction", "none");
+      const std::string turn_direction_2 = lanelet_2.attributeOr("turn_direction", "none");
+      if (turn_direction_1 == turn_direction_2) {
+        continue;
+      }
+
+      if (!hasMergeLane(lanelet_1, lanelet_2, routing_graph_ptr)) {
+        continue;
+      }
+
+      return true;
+    }
+  }
+
+  return false;
+}
+
+}  // namespace
+
+TrafficLightEstimatorNode::TrafficLightEstimatorNode(const rclcpp::NodeOptions & options)
+: Node("traffic_light_estimator", options)
+{
+  using std::placeholders::_1;
+
+  use_last_detect_color_ = this->declare_parameter("use_last_detect_color", true);
```

Comment:
@yukkysaito FYI: If we set `use_last_detect_color_` **true**, the node estimates pedestrian's traffic signal as RED not only when vehicle's traffic signal is detected as GREEN but also when detection results change GREEN to UNKNOWN.

Reply:
when detection results change RED to UNKNOWN, does it estimate as UNKNOWN? only when GREEN to UNKNOW, it is detected as GREEN, right?

Reply:
YES. That's right !

Reply:
I would like to make the name easier to understand. But I have no idea...

Reply:
OK :+1:  Let me think about it ...

Reply:
@yukkysaito I thought about better param name, but I also have no idea... :pleading_face: 
If we describe the parameter's description clearly in the README , I think that we don't have to fix the parameter's name and remain as it is now, what do you think?

I write README in https://github.com/autowarefoundation/autoware.universe/pull/944/commits/138e9c4a8641fb6955bae77191256ce7fb3a5ba2

Reply:
@satoshi-ota Sounds good :+1: